File: 2699-count-the-number-of-fair-pairs/2699-count-the-number-of-fair-pairs.cs
public class Solution {
    public int lower_bound(int [] nums,int i ,int value){
        int indx, low = i, high = nums.Length-1;
        indx = high + 1;
        while(low <= high){
            int mid = (low + high ) >> 1;
            if(nums[mid] >= value){
                high = mid - 1;
                indx = mid;
            }else{
                low = mid + 1;
            }
        }
        return indx;
    }
    public int upper_bound(int [] nums,int value){
        int indx, low = 0, high = nums.Length-1;
        indx = high + 1;
        while(low <= high){
            int mid = (low + high ) >> 1;
            if(nums[mid] <=  value){
                low = mid + 1;
            }else{
                indx = mid;
                high = mid - 1;
            }
        }
        return indx;
    }
    public long CountFairPairs(int[] nums, int lower, int upper) {
        Array.Sort(nums);
        long answer = 0;
        for(int i = 0 ; i < nums.Length; i++){
            int it1 = lower_bound(nums,i+1,lower - nums[i]);
            int it2 = upper_bound(nums, upper-nums[i]);
            if(it2>it1){
                answer += (it2-it1);
            }
        }
        return answer;
    }
}

File: 2699-count-the-number-of-fair-pairs/2699-count-the-number-of-fair-pairs.cpp
class Solution {
public:
    long long countFairPairs(vector<int>& nums, int lower, int upper) {
        sort(nums.begin(),nums.end());
     
        long long ans = 0 ;
         for(int i = 0 ; i < nums.size();i++){
            long long int value1 = lower - nums[i];
            auto it1 = lower_bound(nums.begin()+i+1,nums.end(),value1) - nums.begin();
            int value2 = upper - nums[i];
            auto it2 = upper_bound(nums.begin(),nums.end(),value2) - nums.begin();
            if(it2 >= it1)
            ans += (it2 - it1);
                
            
           
         }
         return ans;
    }
};